src/Aspire.Hosting.Azure.ContainerApps/AzureContanierAppsInfrastructure.cs

```diff
@@ -0,0 +1,991 @@
+
+                    var mapping = context._endpointMapping[epExpr.Endpoint.EndpointName];
+
+                    var val = GetValue(mapping, epExpr.Property);
+
+                    return (val, secretType);
+                }
+
+                if (value is ReferenceExpression expr)
+                {
+                    var args = new object[expr.ValueProviders.Count];
+                    var index = 0;
+                    var finalSecretType = SecretType.None;
+
+                    foreach (var vp in expr.ValueProviders)
+                    {
+                        var (val, secret) = await ProcessValueAsync(vp, executionContext, cancellationToken, secretType).ConfigureAwait(false);
+
+                        // Special case references to keyvault secrets
+                        if (expr.Format == "{0}" && expr.ValueProviders.Count == 1 && secret == SecretType.KeyVault)
+                        {
+                            return (val, secret);
+                        }
+
+                        if (secret != SecretType.None)
+                        {
+                            finalSecretType = SecretType.Normal;
+                        }
+
+                        args[index++] = val;
+                    }
+
+                    return (new BicepValueFormattableString(expr.Format, args), finalSecretType);
+
+                }
+
+                throw new NotSupportedException("Unsupported value type " + value.GetType());
+            }
+
+            private BicepParameter AllocateVolumeStorageAccount(ContainerMountType type, string volumeIndex)
+            {
+                return AllocateParameter(VolumeStorageExpression.GetVolumeStorage(resource, type, volumeIndex));
+            }
+
+            private BicepParameter AllocateContainerImageParameter()
+                => AllocateParameter(ProjectResourceExpression.GetContainerImageExpression((ProjectResource)resource));
+
+            private BicepValue<int> AllocateTargetPortParameter()
+                => AllocateParameter(ProjectResourceExpression.GetTargetPortExpression((ProjectResource)resource));
+
+            private BicepParameter AllocateManagedIdentityIdParameter()
+                => _managedIdentityIdParameter ??= AllocateParameter(_containerAppEnviromentContext.ManagedIdentityId);
+
+            private void AllocateContainerRegistryParameters()
+            {
+                _containerRegistryUrlParameter ??= AllocateParameter(_containerAppEnviromentContext.ContainerRegistryUrl);
+                _containerRegistryManagedIdentityIdParameter ??= AllocateParameter(_containerAppEnviromentContext.ContainerRegistryManagedIdentityId);
+            }
+
+            private BicepParameter AllocateParameter(IManifestExpressionProvider parameter, Type? type = null)
+            {
+                if (!_allocatedParameters.TryGetValue(parameter, out var parameterName))
```

Comment:
Are these `parameter`s always going to be the same instance? We are keying this dictionary on `object`, so it is going to use references.

Reply:
It doesn't matter right now. This is why I was asking @tg-msft about the importance of references and tracking. I track today but it isn't relevant.

Reply:
> It doesn't matter right now.

How doesn't it? if you tried to look up the same parameter with a different instance, it wouldn't match

Reply:
Ah I see what you mean, we do pass the same instance of aspire resources here. That's how we track the names.